aquadoggo/src/db/stores/query.rs

```diff
@@ -0,0 +1,1703 @@
+        let where_ = where_sql(schema.id(), &application_fields, list);
+        let and_filters = where_filter_sql(&args.filter, schema);
+
+        let count_sql = format!(
+            r#"
+            SELECT
+                COUNT(documents.document_id)
+
+            FROM
+                {from}
+
+                JOIN operation_fields_v1
+                    ON
+                        document_view_fields.operation_id = operation_fields_v1.operation_id
+                        AND
+                            document_view_fields.name = operation_fields_v1.name
+
+            WHERE
+                {where_}
+                {and_filters}
+
+            -- Group application fields by name to make sure we get actual number of documents
+            GROUP BY operation_fields_v1.name
+            "#
+        );
+
+        let result: Option<(i64,)> = query_as(&count_sql)
+            .fetch_optional(&self.pool)
+            .await
+            .map_err(|err| DocumentStorageError::FatalStorageError(err.to_string()))?;
+
+        let count = match result {
+            Some(result) => result.0 as u64,
+            None => 0,
+        };
+
+        Ok(count)
+    }
+
+    /// Merges all operation fields from the database into documents.
+    ///
+    /// Due to the special table layout we receive one row per operation field in the query.
+    /// Usually we need to merge multiple rows / operation fields fields into one document.
+    fn convert_rows(
+        rows: Vec<QueryRow>,
+        list: Option<&RelationList>,
+        fields: &ApplicationFields,
+        schema_id: &SchemaId,
+    ) -> Vec<(DocumentCursor, StorageDocument)> {
+        let mut converted: Vec<(DocumentCursor, StorageDocument)> = Vec::new();
+
+        if rows.is_empty() {
+            return converted;
+        }
+
+        // Helper method to convert database row into final document and cursor type
+        let finalize_document = |row: &QueryRow,
+                                 collected_fields: Vec<DocumentViewFieldRow>|
+         -> (DocumentCursor, StorageDocument) {
+            let fields = parse_document_view_field_rows(collected_fields);
+
+            let document = StorageDocument {
+                id: row.document_id.parse().unwrap(),
+                fields: Some(fields),
+                schema_id: schema_id.clone(),
+                view_id: row.document_view_id.parse().unwrap(),
+                author: (&row.owner).into(),
+                deleted: row.is_deleted,
+            };
+
+            let cursor = Self::row_to_cursor(row, list);
+
+            (cursor, document)
+        };
+
+        // Iterate over all database rows and collect fields for each document
+        let last_row = rows.last().unwrap().clone();
+
+        let mut current = rows[0].clone();
+        let mut current_fields = Vec::new();
+
+        let rows_per_document = std::cmp::max(fields.len(), 1);
+
+        for (index, row) in rows.into_iter().enumerate() {
```

Comment:
Does this work because both the `rows` and `fields` are sorted by `document_view_id`?

Reply:
Exactly! I'll verify that again as I did some changes recently which might affect this